src/login/pam_systemd.c

```diff
@@ -219,15 +223,33 @@ static int socket_from_display(const char *display, char **path) {
         memcpy(c, display+1, k);
         c[k] = 0;
 
-        *path = f;
+        r = sockaddr_un_set_path(&sa.un, f);
+        if (r < 0)
+                return r;
+        sa_len = r;
 
-        return 0;
+        fd = socket(AF_UNIX, SOCK_STREAM|SOCK_CLOEXEC, 0);
+        if (fd < 0)
+                return -errno;
+
+        r = connect(fd, &sa.sa, sa_len);
+        if (r >= 0)
+                return TAKE_FD(fd);
+
+        if (errno == ENOENT) {
+                /* Try also abstract socket. We don't need to adjust sa_len since the trailing NUL isn't
```

Comment:
Would it be worth trying the other way around, abstract sock first? There might be other errors other than ENOENT when accessing a filesystem path

Reply:
Makes sense.

Reply:
updated